R/model.R

```diff
@@ -0,0 +1,175 @@
+#'
+#' where:
+#' - \eqn{y} is the predicted value, MFI in our case,
+#' - \eqn{x} is the independent variable, dilution in our case,
+#' - \eqn{B} is the bottom plateau - the right horizontal asymptote,
+#' - \eqn{T} is the top plateau - the left horizontal asymptote,
+#' - \eqn{b} is the slope of the curve at the inflection point,
+#' - \eqn{x_{mid}}{x_mid} is the x-coordinate at the inflection point,
+#' - \eqn{s} is the asymmetric coefficient.
+#'
+#' This equation is refereed as the Richards' equation. More information about the model can be found in the `nplr` package documentation.
+#'
+#' @import nplr
+#' @import dplyr
+#'
+SCModel <- R6::R6Class(
+  "SCModel",
+  public = list(
+    dilutions = NULL,
+    mfi = NULL,
+    mfi_min = NULL,
+    mfi_max = NULL,
+    model = NULL,
+    log_dilution = TRUE,
+    log_mfi = TRUE,
+    scale_mfi = TRUE,
+    # ------------
+    initialize = function(dilutions, mfi, npars = 5, verbose = TRUE, log_dilution = TRUE, log_mfi = TRUE, scale_mfi = TRUE) {
+      stopifnot(length(dilutions) == length(mfi))
+      stopifnot(all((dilutions > 0) & (dilutions < 1)))
+      stopifnot(all(mfi > 0))
+
+      self$log_mfi <- log_mfi
+      self$scale_mfi <- scale_mfi
+      self$log_dilution <- log_dilution
+
+      number_of_samples <- length(dilutions)
+      if (number_of_samples < 5) {
+        verbose_cat(
+          "(", color_codes$red_start, "WARNING", color_codes$red_end, ")\n",
+          "Using less than 5 samples to fit logistic model. For now using the basic nplr method to fit the logistic model - should be modified in the future",
+          verbose = verbose
+        )
+        npars <- min(npars, number_of_samples)
+      }
+
+      mfi <- private$mfi_fit_transform(mfi)
+      self$model <- nplr::nplr(
+        x = dilutions,
+        y = mfi,
+        npars = npars,
+        silent = !verbose,
+        useLog = log_dilution
+      )
+    },
+    #'
+    predict = function(mfi) {
+      if (is.null(self$model)) {
+        stop("SCModel class was not properly initialized. Missing model")
+      }
+      original_mfi <- mfi
+      mfi <- private$mfi_transform(mfi)
+
+      # Example columns: y, x.025, x, x.975
+      df <- nplr::getEstimates(self$model, mfi)
+      # nprl automatically scales the x to non log scale
+      df[, "y"] <- original_mfi
+
+      colnames(df) <- sub("^x", "dilution", colnames(df))
+      colnames(df) <- sub("^y", "MFI", colnames(df))
+      df
+    },
+    plot_data = function() {
+      if (is.null(self$model)) {
+        stop("SCModel class was not properly initialized. Missing model")
+      }
+      targets <- seq(.99, .01, by = -0.01)
+      df <- nplr::getEstimates(self$model, targets = targets)
+      df[, "y"] <- private$mfi_reverse_transform(df[, "y"])
+      colnames(df) <- sub("^x", "dilution", colnames(df))
+      colnames(df) <- sub("^y", "MFI", colnames(df))
+      df
+    },
+
+    #'
+    top_asymptote = function() {
```

Comment:
I would move this into the `active` part of the class, it is not a typical method, but rather a field. Also missing docs